packages/filesystem/src/electron-browser/file-dialog/electron-file-dialog-service.ts

```diff
@@ -0,0 +1,165 @@
+/********************************************************************************
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable } from 'inversify';
+import { remote, FileFilter, OpenDialogOptions, SaveDialogOptions } from 'electron';
+import URI from '@theia/core/lib/common/uri';
+import { isOSX } from '@theia/core/lib/common/os';
+import { MaybeArray } from '@theia/core/lib/common/types';
+import { FileAccess } from '@theia/filesystem/lib/common/filesystem';
+import { MessageService } from '@theia/core/lib/common/message-service';
+import { FileStat } from '../../common';
+import { DefaultFileDialogService, OpenFileDialogProps, SaveFileDialogProps } from '../../browser/file-dialog';
+
+import { FileUri } from '@theia/core/lib/node/file-uri'; // We are OK to use this here.
```

Comment:
TODO: document this!

Reply:
```
//
// We are OK to use this here because the electron backend and frontend are on the same host.
// If required, we can move this single service (and its module) to a dedicated Theia extension,
// and at packaging time, clients can decide whether they need the native or the browser-based
// solution.
//
```